# Accurate Annotation of Protein Sequences with IDTAXA

**Table of contents:**
- [Introduction](#introduction)
- [Requirements](#requirements)
- [Usage](#usage)
- [Outputs](#Outputs)
- [Reporting Bugs and Errors](#Reporting-Bugs-and-Errors)

## Introduction

This repository contains scripts for reproducing figures in the manuscript Accurate Annotation of Protein Sequences with IDTAXA. For the sake of reproducibility, these scripts do not assume any particular parallelization schemes, or schedulers. Processes are run sequentially within scripts to make them as broadly accessible as possible. These scripts should work in both Linux or Mac environments, and have not been tested on Windows.

These scripts are designed to work with input files that are present on zenodo. Outputs of these scripts used in the construction of figures are also present on zenodo.

## Requirements

These scripts have a few minimal requirements:

1. R 4.0.0 or greater
2. DECIPHER 2.19.0 or greater
3. BLAST 2.10.1 or greater
4. HMMER 3.3.1 or greater
5. R must have access to both HMMER and BLAST
6. gridExtra is required to plot tables

## Usage

4 RScripts are provided in this repo, they should be run directly through the `Rscript` command in terminal and as long as R has access to HMMER and BLAST when called upon in this way, and the required packages are installed, they should run to completion. For reference the approximate run times of the three main scripts on a 2019 Macbook Pro with 6 cores and 64 GB of RAM are on R 4.1.0, with HMMER 3.3.1 and BLAST 2.10.1:

1. `ICrossVal.R` ~ 6 hours
2. `BCrossVal.R` ~ for AAs 12-18 hours, for NTs 2-10 minutes (sw_traceback is used for AAs)
3. `HCrossVal.R` ~ 48 hours+ - a large portion of this is model construction

These times are meant to serve only as a guidepost, your own mileage may vary.

The script `ErrorPlot.R` will take in results generated by any of these scripts and plot them in a manner similar to what is present in the manuscript.

### ICrossVal.R

The script `ICrossVal.R` takes in training data that consists of sequences, taxonomies, and holdout designations and saves off an RData file that contains the objects `IPerf` which details the performance of IDTaxa on the training set, and `IRes` the raw result of IDTaxa's cross validations.

The script requires three trailing arguments, and infers a fourth if it is not supplied.

```
$ Rscript </path/to/script/ICrossVal.R> <Input.RData> <1> <YourWellNamedOutput.RData>
```

Like it's three counterparts `ICrossVal.R` expects input data to be a very specific format. The training data used in the paper is present in a zenodo repository HERE, but if users can replicate this format they could certainly supply their own. The integer argument references to the construction of the input training data and for the supplied data used in the paper can be 1-5. The third argument is simply the name users wish to assign to their output file. The fourth inferred argument allows users to force IDTAXA to use the standard amino acid alphabet, by simply including `TRUE` as a fourth argument. If it is not included it is inferred as `FALSE`. This script would require manipulation to produce the result for the literature alphabet supplied in the paper, but will work with the nucleotide data just fine.

### BCrossVal.R

This script takes in the same three major arguments and assumes others if they are not supplied. Extra arguments are explained in the comments of the script. It has a similar usage.

```
$ Rscript </path/to/script/BCrossVal.R> <Input.RData> <1> <YourWellNamedOutput.RData>
```

### HCrossVal.R

This script takes in the same three major arguments and assumes others if they are not supplied. Extra arguments are explained in the comments of the script. It has a similar usage.

```
$ Rscript </path/to/script/HCrossVal.R> <Input.RData> <1> <YourWellNamedOutput.RData>
```

These scripts were written to be minimally reproducible, they are not efficient computing. They do not assume anything beyond base R, DECIPHER v 2.19.0 (or greater), and the other tools that fundamentally underlie competing classification strategies. All of these scripts can be sped up considerably by any number of parallelization regimes.

### ErrorPlot.R

This script takes in results generated by the previous three scripts and plots them. It has a unique argument syntax that is not particularly user friendly, but does work. Users supply the filepaths to results generated by the preceding scripts and seperate those results with SEP where they wish to delineate a new panel. not including any SEP's will lead to all supplied results being plotted in a single panel. The script is modestly smart, but not written to accommodate an adversarial user. It can generate a 1, 2, 3, or 4 panel figure, and will attempt to provide a legend to each panel with names derived from the input file names. It will additionally attempt to provide tables describing the Over-classification and Miss-classification rates present in the supplied data fixed at 60% confidence for each method, and fixed at 60% of non-singletons classified for each method. The script expects IDTAXA results to have names in the form of `METHOD_TaxLevel_AdditionalDescription.RData`, and HMMER and BLAST results to have names in the form of `METHOD_TaxLevel.RDATA`. Legend names and table entries will be derived and organized from these names. Data that are grouped together will be placed in the same panel.

```
$ Rscript </path/to/script/ErrorPlot.R> <ResultFile_A_1.RDATA> <SEP> <ResultFile_A_2.RData>
```

## Outputs

The cross validation scripts take in data with a specific format, and return data of data with a specific format. Users can either run these with the supplied data used in the paper, or generate their own.

The plotting script returns plots similar to, but not exact replicas of those present in the paper. Example outputs of this plotting script are provided below:

![Error Rates Plotted](https://github.com/npcooley/AAClassification/blob/master/errorplot.pdf)

![Error Rates when classifiers are 60% Confident](https://github.com/npcooley/AAClassification/blob/master/ErrByConf60.pdf)

![Error Rates when classifiers have classified 60% of classifiable sequences](https://github.com/npcooley/AAClassification/blob/master/ErrByClass60.pdf)

These output files have hard coded output names as `ErrorPlot.pdf`, `ErrByClass60.pdf`, and `ErrByConf60.pdf` respectively, and will be placed in R's working directory.

## Reporting Bugs and Errors

Please feel free to contact me with any concerns, questions, or bugs at npc19@pitt.edu, or submit issues here on Github.


